### DISCUSSION TITLE
How to prevent students from installing packages?

### DISCUSSION THREAD

**Discussion by abdullahkhalids** [jupyter/nbgrader#1880] [Q&A] [answered]
My students will often put installation instructions (via cells starting with `!pip install` )in their submissions. This results in these packages installed on the "instructor computer" when the grader runs on the submission. Is there any way, I can ask nbgrader to ignore these cells?

**Comment by jeflem** [comment]
In your grading code you may use the variable `_i` to access the source code of the previous cell (the student's answer). Search it for unwanted strings like `!pip` or `import xyz` (to prohibit using package xyz). See [IPython doc](https://ipython.readthedocs.io/en/stable/interactive/reference.html#input-caching-system) for more such variables.

**Comment by abdullahkhalids** [comment, marked as answer]
My solution was to remove the pip module itself from the virtual environment. I did that by just renaming the pip folder inside the virtual environment.
